xtask/src/release.rs

```diff
@@ -1,67 +1,147 @@
-use std::path::Path;
-
+use std::{
+    io::{stdin, stdout, BufRead, Write},
+    path::{Path, PathBuf},
+};
+
+use isahc::{
+    auth::{Authentication, Credentials},
+    config::Configurable,
+    http::StatusCode,
+    HttpClient, ReadResponseExt, Request,
+};
 use itertools::Itertools;
+use semver::Version;
 use serde::Deserialize;
-use serde_json::json;
+use serde_json::{json, Value as JsonValue};
 use toml::from_str as from_toml_str;
-use xshell::{cmd, pushd, read_file};
+use xshell::{pushd, read_file};
 
-use crate::{config, flags, project_root, Result};
+use crate::{cmd, config, Result};
 
-const GITHUB_API_RELEASES: &str = "https://api.github.com/repos/ruma/ruma/releases";
+const CRATESIO_API: &str = "https://crates.io/api/v1/crates";
+const GITHUB_API_RUMA: &str = "https://api.github.com/repos/zecakeh/ruma";
 
-impl flags::Release {
-    /// Run the release command to effectively create a release.
-    pub(crate) fn run(self) -> Result<()> {
-        let project_root = &project_root()?;
-        let _dir = pushd(project_root.join(&self.name))?;
+/// Task to create a new release of the given crate.
+pub struct ReleaseTask {
+    /// The crate to release.
+    name: String,
+
+    /// The root of the workspace.
+    project_root: PathBuf,
+
+    /// The version to release.
+    version: Version,
+
+    /// The http client to use for requests.
+    client: HttpClient,
+}
+
+impl ReleaseTask {
+    /// Create a new `ReleaseTask` with the given `name` and `project_root`.
+    pub(crate) fn new(name: String, project_root: PathBuf) -> Result<Self> {
+        let path = project_root.join(&name);
 
-        let remote = &self.get_remote()?;
+        let version = Self::get_version(&path)?;
 
-        if !cmd!("git status -s -uno").read()?.is_empty() {
-            return Err("This git repository contains untracked files".into());
+        Ok(Self { name, project_root, version, client: HttpClient::new()? })
+    }
+
+    /// Run the task to effectively create a release.
+    pub(crate) fn run(self) -> Result<()> {
+        println!("Starting release for {} {}…", self.name, self.version);
+
+        if self.is_released()? {
+            return Err("This version is already released".into());
         }
 
-        let version = &self.get_version(project_root)?;
-        println!("Making release for {} {}…", self.name, version);
+        let _dir = pushd(self.get_path())?;
+
+        let remote = Self::get_remote()?;
+
+        // println!("Checking status of git repository…");
+        // if !cmd!("git status -s -uno").read()?.is_empty() {
+        //     return Err("This git repository contains untracked files".into());
+        // }
```

Comment:
Maybe this should be made into a warning, and generate another "are you sure" prompt, then one doesn't have to comment it out so much (I also did that a few times).